File: Algorithms/CP/Haskell/cp.hs
{-# LANGUAGE BangPatterns #-}
module Main where

import System.Environment (getArgs)
import System.IO (readFile)
import Control.Monad.ST
import Control.Monad (when, forM, forM_)
import Data.STRef
import Data.Array.ST
import Data.Bits
import Text.Printf (printf)
import Data.List (minimumBy)
import Data.Ord (comparing)

-- Type aliases
type Grid s = STUArray s (Int, Int) Int
type Candidates s = STUArray s (Int, Int) Int

-- Parse matrix file
parseMatrix :: String -> [[Int]]
parseMatrix content =
    let validLines = filter (\l -> not (null l) && head l /= '#') (lines content)
        rows = take 9 validLines
    in map (map (\w -> read w :: Int) . words) rows

-- Print board formatted
printBoard :: [[Int]] -> IO ()
printBoard board = do
    putStrLn "\nPuzzle:"
    mapM_ (\row -> putStrLn $ unwords (map show row) ++ " ") board

-- Main entry point
main :: IO ()
main = do
    args <- getArgs
    let matrixFiles = filter (\f -> reverse (take 7 (reverse f)) == ".matrix") args
    mapM_ solveFile matrixFiles

solveFile :: String -> IO ()
solveFile filename = do
    putStrLn filename
    content <- readFile filename
    let board = parseMatrix content
    -- Print raw board
    mapM_ (\row -> putStrLn $ unwords (map show row) ++ " ") board
    printBoard board

    let result = runST $ solveCP board
    case result of
        Just (solution, iters) -> do
            printBoard solution
            printf "\nSolved in Iterations=%d\n\n" iters
        Nothing -> putStrLn "No solution found"

-- Solve Sudoku using Constraint Propagation
solveCP :: [[Int]] -> ST s (Maybe ([[Int]], Int))
solveCP puzzle = do
    -- Create mutable grid and candidates arrays
    grid <- newArray ((0,0), (8,8)) 0 :: ST s (STUArray s (Int, Int) Int)
    candidates <- newArray ((0,0), (8,8)) 0 :: ST s (STUArray s (Int, Int) Int)

    -- Create iteration counter
    iterRef <- newSTRef 0

    -- Initialize grid and candidates
    initGrid grid candidates puzzle iterRef

    -- Propagate initial constraints
    success <- propagate grid candidates iterRef

    if not success
        then return Nothing
        else do
            -- Run search
            found <- cpSearch grid candidates iterRef

            if found
                then do
                    solution <- extractSolution grid
                    iters <- readSTRef iterRef
                    return $ Just (solution, iters)
                else return Nothing

-- Initialize grid and candidates
initGrid :: Grid s -> Candidates s -> [[Int]] -> STRef s Int -> ST s ()
initGrid grid candidates puzzle iterRef = do
    forM_ [0..8] $ \r -> do
        forM_ [0..8] $ \c -> do
            let val = (puzzle !! r) !! c
            if val == 0
                then do
                    -- Empty cell: all candidates 1-9
                    writeArray grid (r, c) 0
                    writeArray candidates (r, c) 0x3FE  -- bits 1-9 set
                else do
                    -- Given clue
                    writeArray grid (r, c) val
                    writeArray candidates (r, c) (shiftL 1 val)

-- Get all peers for a cell (20 cells: 8 in row + 8 in col + 4 in box)
getPeers :: Int -> Int -> [(Int, Int)]
getPeers row col =
    let rowPeers = [(row, c) | c <- [0..8], c /= col]
        colPeers = [(r, col) | r <- [0..8], r /= row]
        boxRow = (row `div` 3) * 3
        boxCol = (col `div` 3) * 3
        boxPeers = [(r, c) | r <- [boxRow..boxRow+2], c <- [boxCol..boxCol+2], r /= row, c /= col]
    in rowPeers ++ colPeers ++ boxPeers

-- Check if digit is a candidate
hasCandidate :: Int -> Int -> Bool
hasCandidate cands digit = testBit cands digit

-- Remove digit from candidates
removeCandidate :: Int -> Int -> Int
removeCandidate cands digit = clearBit cands digit

-- Count number of candidates
countCandidates :: Int -> Int
countCandidates cands = popCount cands

-- Get first candidate digit
getFirstCandidate :: Int -> Maybe Int
getFirstCandidate cands =
    case [d | d <- [1..9], testBit cands d] of
        [] -> Nothing
        (d:_) -> Just d

-- Eliminate a digit from a cell
eliminate :: Grid s -> Candidates s -> STRef s Int -> Int -> Int -> Int -> ST s Bool
eliminate grid candidates iterRef row col digit = do
    cands <- readArray candidates (row, col)

    -- Already eliminated
    if not (hasCandidate cands digit)
        then return True
        else do
            -- Remove digit
            let newCands = removeCandidate cands digit
            writeArray candidates (row, col) newCands

            -- Check for contradiction
            let remaining = countCandidates newCands
            if remaining == 0
                then return False
                else if remaining == 1
                    then do
                        -- Only one candidate left - assign it
                        val <- readArray grid (row, col)
                        if val == 0
                            then case getFirstCandidate newCands of
                                Just lastDigit -> assign grid candidates iterRef row col lastDigit
                                Nothing -> return False
                            else return True
                    else return True

-- Assign a digit to a cell
assign :: Grid s -> Candidates s -> STRef s Int -> Int -> Int -> Int -> ST s Bool
assign grid candidates iterRef row col digit = do
    -- Check if already assigned (avoid double-counting)
    currentVal <- readArray grid (row, col)
    if currentVal /= 0
        then return (currentVal == digit)  -- OK if same digit, contradiction if different
        else do
            -- Increment iteration counter
            modifySTRef iterRef (+1)

            -- Set value
            writeArray grid (row, col) digit
            writeArray candidates (row, col) (shiftL 1 digit)

            -- Eliminate digit from all peers
            let peers = getPeers row col
            let eliminateFromPeer (pr, pc) = eliminate grid candidates iterRef pr pc digit

            results <- mapM eliminateFromPeer peers
            return $ and results

-- Propagate constraints (incremental like C reference)
propagate :: Grid s -> Candidates s -> STRef s Int -> ST s Bool
propagate grid candidates iterRef = propagateLoop
  where
    propagateLoop = do
        -- Strategy 1: Singleton elimination (process all, track changes)
        (success1, changed1) <- processSingletons
        if not success1
            then return False
            else do
                -- Strategy 2: Hidden singles in rows
                (success2, changed2) <- processHiddenSinglesRows
                if not success2
                    then return False
                    else do
                        -- Strategy 3: Hidden singles in columns
                        (success3, changed3) <- processHiddenSinglesCols
                        if not success3
                            then return False
                            else do
                                -- Strategy 4: Hidden singles in boxes
                                (success4, changed4) <- processHiddenSinglesBoxes
                                if not success4
                                    then return False
                                    else do
                                        let anyChanged = changed1 || changed2 || changed3 || changed4
                                        if anyChanged
                                            then propagateLoop
                                            else return True

    -- Process singletons incrementally
    processSingletons = processCells 0 0 False
      where
        processCells r c changed
            | r >= 9 = return (True, changed)
            | c >= 9 = processCells (r + 1) 0 changed
            | otherwise = do
                val <- readArray grid (r, c)
                if val == 0
                    then do
                        cands <- readArray candidates (r, c)
                        let numCands = countCandidates cands
                        if numCands == 0
                            then return (False, changed)  -- Contradiction
                            else if numCands == 1
                                then case getFirstCandidate cands of
                                    Just digit -> do
                                        success <- assign grid candidates iterRef r c digit
                                        if success
                                            then processCells r (c + 1) True
                                            else return (False, changed)
                                    Nothing -> return (False, changed)
                                else processCells r (c + 1) changed
                    else processCells r (c + 1) changed

    -- Process hidden singles in rows incrementally
    processHiddenSinglesRows = processRows 0 False
      where
        processRows r changed
            | r >= 9 = return (True, changed)
            | otherwise = do
                (success, rowChanged) <- processDigits r 1 False
                if success
                    then processRows (r + 1) (changed || rowChanged)
                    else return (False, changed)

        processDigits r digit changed
            | digit > 9 = return (True, changed)
            | otherwise = do
                -- Check if digit already assigned in row
                rowVals <- mapM (\c -> readArray grid (r, c)) [0..8]
                if digit `elem` rowVals
                    then processDigits r (digit + 1) changed
                    else do
                        -- Count cells where digit is candidate
                        (count, lastCol) <- countInRow r digit 0 0 (-1)
                        if count == 1
                            then do
                                success <- assign grid candidates iterRef r lastCol digit
                                if success
                                    then processDigits r (digit + 1) True
                                    else return (False, changed)
                            else if count == 0
                                then return (False, changed)  -- Contradiction
                                else processDigits r (digit + 1) changed

        countInRow r digit c count lastCol
            | c >= 9 = return (count, lastCol)
            | otherwise = do
                cands <- readArray candidates (r, c)
                if hasCandidate cands digit
                    then countInRow r digit (c + 1) (count + 1) c
                    else countInRow r digit (c + 1) count lastCol

    -- Process hidden singles in columns incrementally
    processHiddenSinglesCols = processCols 0 False
      where
        processCols c changed
            | c >= 9 = return (True, changed)
            | otherwise = do
                (success, colChanged) <- processDigits c 1 False
                if success
                    then processCols (c + 1) (changed || colChanged)
                    else return (False, changed)

        processDigits c digit changed
            | digit > 9 = return (True, changed)
            | otherwise = do
                -- Check if digit already assigned in column
                colVals <- mapM (\r -> readArray grid (r, c)) [0..8]
                if digit `elem` colVals
                    then processDigits c (digit + 1) changed
                    else do
                        -- Count cells where digit is candidate
                        (count, lastRow) <- countInCol c digit 0 0 (-1)
                        if count == 1
                            then do
                                success <- assign grid candidates iterRef lastRow c digit
                                if success
                                    then processDigits c (digit + 1) True
                                    else return (False, changed)
                            else if count == 0
                                then return (False, changed)  -- Contradiction
                                else processDigits c (digit + 1) changed

        countInCol c digit r count lastRow
            | r >= 9 = return (count, lastRow)
            | otherwise = do
                cands <- readArray candidates (r, c)
                if hasCandidate cands digit
                    then countInCol c digit (r + 1) (count + 1) r
                    else countInCol c digit (r + 1) count lastRow

    -- Process hidden singles in boxes incrementally
    processHiddenSinglesBoxes = processBoxes 0 False
      where
        processBoxes box changed
            | box >= 9 = return (True, changed)
            | otherwise = do
                let boxRow = (box `div` 3) * 3
                let boxCol = (box `mod` 3) * 3
                (success, boxChanged) <- processDigits boxRow boxCol 1 False
                if success
                    then processBoxes (box + 1) (changed || boxChanged)
                    else return (False, changed)

        processDigits boxRow boxCol digit changed
            | digit > 9 = return (True, changed)
            | otherwise = do
                -- Check if digit already assigned in box
                boxVals <- sequence [readArray grid (r, c) | r <- [boxRow..boxRow+2], c <- [boxCol..boxCol+2]]
                if digit `elem` boxVals
                    then processDigits boxRow boxCol (digit + 1) changed
                    else do
                        -- Count cells where digit is candidate
                        (count, lastR, lastC) <- countInBox boxRow boxCol digit 0 0 0 (-1) (-1)
                        if count == 1
                            then do
                                success <- assign grid candidates iterRef lastR lastC digit
                                if success
                                    then processDigits boxRow boxCol (digit + 1) True
                                    else return (False, changed)
                            else if count == 0
                                then return (False, changed)  -- Contradiction
                                else processDigits boxRow boxCol (digit + 1) changed

        countInBox boxRow boxCol digit dr dc count lastR lastC
            | dr >= 3 = return (count, lastR, lastC)
            | dc >= 3 = countInBox boxRow boxCol digit (dr + 1) 0 count lastR lastC
            | otherwise = do
                let r = boxRow + dr
                let c = boxCol + dc
                cands <- readArray candidates (r, c)
                if hasCandidate cands digit
                    then countInBox boxRow boxCol digit dr (dc + 1) (count + 1) r c
                    else countInBox boxRow boxCol digit dr (dc + 1) count lastR lastC

-- Find MRV cell (cell with minimum remaining values)
findMRVCell :: Grid s -> Candidates s -> ST s (Maybe (Int, Int))
findMRVCell grid candidates = do
    cells <- forM [0..8] $ \r -> do
        forM [0..8] $ \c -> do
            val <- readArray grid (r, c)
            if val == 0
                then do
                    cands <- readArray candidates (r, c)
                    let numCands = countCandidates cands
                    return [(r, c, numCands)]
                else return []
    let emptyCells = concat $ concat cells
    if null emptyCells
        then return Nothing
        else do
            let (r, c, _) = minimumBy (comparing (\(_,_,n) -> n)) emptyCells
            return $ Just (r, c)

-- CP search with backtracking
cpSearch :: Grid s -> Candidates s -> STRef s Int -> ST s Bool
cpSearch grid candidates iterRef = do
    -- Find MRV cell
    maybeCell <- findMRVCell grid candidates
    case maybeCell of
        Nothing -> return True  -- No empty cells - solved!
        Just (r, c) -> do
            -- Get candidates for this cell
            cands <- readArray candidates (r, c)
            let digits = [d | d <- [1..9], hasCandidate cands d]

            -- Try each candidate
            tryDigits digits r c

  where
    tryDigits [] _ _ = return False
    tryDigits (d:ds) r c = do
        -- Save state
        gridCopy <- copyGrid grid
        candsCopy <- copyCandidates candidates

        -- Try assigning digit
        success <- assign grid candidates iterRef r c d
        if success
            then do
                -- Propagate and recurse
                propSuccess <- propagate grid candidates iterRef
                if propSuccess
                    then do
                        found <- cpSearch grid candidates iterRef
                        if found
                            then return True
                            else do
                                -- Backtrack
                                restoreGrid grid gridCopy
                                restoreCandidates candidates candsCopy
                                tryDigits ds r c
                    else do
                        -- Backtrack
                        restoreGrid grid gridCopy
                        restoreCandidates candidates candsCopy
                        tryDigits ds r c
            else do
                -- Backtrack
                restoreGrid grid gridCopy
                restoreCandidates candidates candsCopy
                tryDigits ds r c

-- Copy grid
copyGrid :: Grid s -> ST s [[Int]]
copyGrid grid = do
    rows <- forM [0..8] $ \r -> do
        forM [0..8] $ \c -> do
            readArray grid (r, c)
    return rows

-- Copy candidates
copyCandidates :: Candidates s -> ST s [[Int]]
copyCandidates cands = do
    rows <- forM [0..8] $ \r -> do
        forM [0..8] $ \c -> do
            readArray cands (r, c)
    return rows

-- Restore grid
restoreGrid :: Grid s -> [[Int]] -> ST s ()
restoreGrid grid gridCopy = do
    forM_ [0..8] $ \r -> do
        forM_ [0..8] $ \c -> do
            writeArray grid (r, c) ((gridCopy !! r) !! c)

-- Restore candidates
restoreCandidates :: Candidates s -> [[Int]] -> ST s ()
restoreCandidates cands candsCopy = do
    forM_ [0..8] $ \r -> do
        forM_ [0..8] $ \c -> do
            writeArray cands (r, c) ((candsCopy !! r) !! c)

-- Extract solution from grid
extractSolution :: Grid s -> ST s [[Int]]
extractSolution grid = do
    forM [0..8] $ \r -> do
        forM [0..8] $ \c -> do
            readArray grid (r, c)

File: Algorithms/CP/Erlang/cp.erl
#!/usr/bin/env escript
%% -*- erlang -*-
%% Erlang Constraint Propagation (CP) Sudoku Solver
%% Port of C CP implementation using arrays and bitsets

main([Filename|_]) ->
    %% Initialize iteration counter using process dictionary
    put(cp_iterations, 0),

    %% Read puzzle
    Puzzle = read_matrix(Filename),
    print_puzzle("Puzzle", Puzzle),

    %% Initialize grid
    {Grid, Candidates} = init_grid(Puzzle),

    %% Solve using CP
    case solve_cp(Grid, Candidates) of
        {ok, Solution} ->
            print_puzzle("Puzzle", Solution),
            io:format("~nSolved in Iterations=~B~n~n", [get(cp_iterations)]);
        error ->
            io:format("~nNo solution found~n")
    end;
main(_) ->
    io:format(standard_error, "Usage: escript cp.erl <matrix_file>~n", []),
    halt(1).

%% Read matrix from file
read_matrix(Filename) ->
    %% Print filename (normalize /app/Matrices to ../Matrices)
    case string:prefix(Filename, "/app/Matrices/") of
        nomatch -> io:format("~s~n", [Filename]);
        Rest -> io:format("../Matrices/~s~n", [Rest])
    end,
    {ok, Binary} = file:read_file(Filename),
    Lines = string:split(binary_to_list(Binary), "\n", all),
    Grid = array:new([{size, 9}, {default, array:new([{size, 9}, {default, 0}])}]),
    read_rows(Lines, 0, Grid).

read_rows(_, 9, Grid) -> Grid;
read_rows([], _, Grid) -> Grid;
read_rows([Line|Rest], Row, Grid) ->
    Trimmed = string:trim(Line),
    case Trimmed of
        "" -> read_rows(Rest, Row, Grid);
        [$#|_] -> read_rows(Rest, Row, Grid);  %% Skip comments
        _ ->
            Numbers = string:tokens(Trimmed, " "),
            NewGrid = parse_row(Numbers, Row, 0, Grid),
            io:format("~n"),
            read_rows(Rest, Row + 1, NewGrid)
    end.

parse_row([], _, _, Grid) -> Grid;
parse_row([NumStr|Rest], Row, Col, Grid) when Col < 9 ->
    Num = list_to_integer(NumStr),
    io:format("~B ", [Num]),
    RowArray = array:get(Row, Grid),
    NewRowArray = array:set(Col, Num, RowArray),
    NewGrid = array:set(Row, NewRowArray, Grid),
    parse_row(Rest, Row, Col + 1, NewGrid);
parse_row(_, _, _, Grid) -> Grid.

%% Print puzzle
print_puzzle(Label, Grid) ->
    io:format("~n~s:~n", [Label]),
    array:map(fun(_R, RowArray) ->
        array:map(fun(_C, Val) ->
            io:format("~B ", [Val]),
            Val
        end, RowArray),
        io:format("~n"),
        RowArray
    end, Grid),
    Grid.

%% Bitset helpers
has_candidate(Set, Digit) -> (Set band (1 bsl Digit)) =/= 0.

remove_candidate(Set, Digit) -> Set band (bnot (1 bsl Digit)).

%% Count number of set bits (popcount)
count_candidates(Set) ->
    count_bits(Set, 0).

count_bits(0, Acc) -> Acc;
count_bits(N, Acc) -> count_bits(N band (N - 1), Acc + 1).

%% Get first candidate digit from bitset (1-9)
get_first_candidate(Cs) ->
    get_first_candidate(Cs, 1).

get_first_candidate(_, 10) -> 0;
get_first_candidate(Cs, Digit) ->
    case has_candidate(Cs, Digit) of
        true -> Digit;
        false -> get_first_candidate(Cs, Digit + 1)
    end.

%% Get all 20 peers for a cell (row, col, box)
get_peers(Row, Col) ->
    %% Same row (8 cells)
    RowPeers = [{Row, C} || C <- lists:seq(0, 8), C =/= Col],

    %% Same column (8 cells)
    ColPeers = [{R, Col} || R <- lists:seq(0, 8), R =/= Row],

    %% Same 3x3 box (4 cells not already counted)
    BoxRow = (Row div 3) * 3,
    BoxCol = (Col div 3) * 3,
    BoxPeers = [{R, C} || R <- lists:seq(BoxRow, BoxRow + 2),
                          C <- lists:seq(BoxCol, BoxCol + 2),
                          R =/= Row andalso C =/= Col],

    RowPeers ++ ColPeers ++ BoxPeers.

%% Initialize grid from puzzle
init_grid(Puzzle) ->
    Grid = array:new([{size, 9}, {default, array:new([{size, 9}, {default, 0}])}]),
    Candidates = array:new([{size, 9}, {default, array:new([{size, 9}, {default, 0}])}]),

    {NewGrid, NewCandidates} = init_cells(Puzzle, Grid, Candidates, 0, 0),

    %% Initial constraint propagation for given clues
    {NewGrid, NewCandidates}.

init_cells(_, Grid, Candidates, 9, _) -> {Grid, Candidates};
init_cells(Puzzle, Grid, Candidates, Row, 9) ->
    init_cells(Puzzle, Grid, Candidates, Row + 1, 0);
init_cells(Puzzle, Grid, Candidates, Row, Col) ->
    PuzzleRow = array:get(Row, Puzzle),
    Val = array:get(Col, PuzzleRow),

    {NewGrid, NewCandidates} = if
        Val =:= 0 ->
            %% Empty cell: set all candidates 1-9
            GridRow = array:get(Row, Grid),
            NewGridRow = array:set(Col, 0, GridRow),
            G = array:set(Row, NewGridRow, Grid),

            CandRow = array:get(Row, Candidates),
            NewCandRow = array:set(Col, 16#3FE, CandRow),  %% Binary: 0011 1111 1110 (bits 1-9)
            C = array:set(Row, NewCandRow, Candidates),
            {G, C};
        true ->
            %% Given clue: set single value
            GridRow = array:get(Row, Grid),
            NewGridRow = array:set(Col, Val, GridRow),
            G = array:set(Row, NewGridRow, Grid),

            CandRow = array:get(Row, Candidates),
            NewCandRow = array:set(Col, 1 bsl Val, CandRow),
            C = array:set(Row, NewCandRow, Candidates),
            {G, C}
    end,

    init_cells(Puzzle, NewGrid, NewCandidates, Row, Col + 1).

%% Get grid value
get_grid(Grid, Row, Col) ->
    RowArray = array:get(Row, Grid),
    array:get(Col, RowArray).

%% Set grid value
set_grid(Grid, Row, Col, Val) ->
    RowArray = array:get(Row, Grid),
    NewRowArray = array:set(Col, Val, RowArray),
    array:set(Row, NewRowArray, Grid).

%% Get candidates
get_candidates(Candidates, Row, Col) ->
    RowArray = array:get(Row, Candidates),
    array:get(Col, RowArray).

%% Set candidates
set_candidates(Candidates, Row, Col, Val) ->
    RowArray = array:get(Row, Candidates),
    NewRowArray = array:set(Col, Val, RowArray),
    array:set(Row, NewRowArray, Candidates).

%% Eliminate a digit from a cell's candidates
eliminate(Grid, Candidates, Row, Col, Digit) ->
    CandSet = get_candidates(Candidates, Row, Col),

    %% Check if digit is already eliminated
    case has_candidate(CandSet, Digit) of
        false ->
            {ok, Grid, Candidates};  %% Already eliminated
        true ->
            %% Remove digit from candidates
            NewCandSet = remove_candidate(CandSet, Digit),
            NewCandidates = set_candidates(Candidates, Row, Col, NewCandSet),

            %% Check for contradiction
            Remaining = count_candidates(NewCandSet),
            if
                Remaining =:= 0 ->
                    error;  %% Contradiction
                Remaining =:= 1 ->
                    case get_grid(Grid, Row, Col) of
                        0 ->
                            %% Singleton elimination
                            LastDigit = get_first_candidate(NewCandSet),
                            assign(Grid, NewCandidates, Row, Col, LastDigit);
                        _ ->
                            {ok, Grid, NewCandidates}
                    end;
                true ->
                    {ok, Grid, NewCandidates}
            end
    end.

%% Assign a digit to a cell
assign(Grid, Candidates, Row, Col, Digit) ->
    %% Increment iteration counter
    put(cp_iterations, get(cp_iterations) + 1),

    %% Set value
    NewGrid = set_grid(Grid, Row, Col, Digit),
    NewCandidates = set_candidates(Candidates, Row, Col, 1 bsl Digit),

    %% Eliminate digit from all peers
    Peers = get_peers(Row, Col),
    eliminate_from_peers(NewGrid, NewCandidates, Peers, Digit).

eliminate_from_peers(Grid, Candidates, [], _Digit) ->
    {ok, Grid, Candidates};
eliminate_from_peers(Grid, Candidates, [{PeerRow, PeerCol}|Rest], Digit) ->
    case eliminate(Grid, Candidates, PeerRow, PeerCol, Digit) of
        {ok, NewGrid, NewCandidates} ->
            eliminate_from_peers(NewGrid, NewCandidates, Rest, Digit);
        error -> error
    end.

%% Constraint propagation
propagate(Grid, Candidates) ->
    case propagate_loop(Grid, Candidates) of
        {ok, NewGrid, NewCandidates, _Changed} -> {ok, NewGrid, NewCandidates};
        error -> error
    end.

propagate_loop(Grid, Candidates) ->
    %% Strategy 1: Singleton elimination
    case singleton_elimination(Grid, Candidates, 0, 0) of
        {ok, NewGrid, NewCandidates, Changed1} ->
            %% Strategy 2: Hidden singles - rows
            case hidden_singles_rows(NewGrid, NewCandidates, 0, 1) of
                {ok, NewGrid2, NewCandidates2, Changed2} ->
                    %% Strategy 3: Hidden singles - columns
                    case hidden_singles_cols(NewGrid2, NewCandidates2, 0, 1) of
                        {ok, NewGrid3, NewCandidates3, Changed3} ->
                            %% Strategy 4: Hidden singles - boxes
                            case hidden_singles_boxes(NewGrid3, NewCandidates3, 0, 1) of
                                {ok, NewGrid4, NewCandidates4, Changed4} ->
                                    Changed = Changed1 orelse Changed2 orelse Changed3 orelse Changed4,
                                    if
                                        Changed -> propagate_loop(NewGrid4, NewCandidates4);
                                        true -> {ok, NewGrid4, NewCandidates4, false}
                                    end;
                                error -> error
                            end;
                        error -> error
                    end;
                error -> error
            end;
        error -> error
    end.

%% Singleton elimination
singleton_elimination(Grid, Candidates, 9, _) -> {ok, Grid, Candidates, false};
singleton_elimination(Grid, Candidates, Row, 9) ->
    singleton_elimination(Grid, Candidates, Row + 1, 0);
singleton_elimination(Grid, Candidates, Row, Col) ->
    case get_grid(Grid, Row, Col) of
        0 ->
            CandSet = get_candidates(Candidates, Row, Col),
            NumCandidates = count_candidates(CandSet),
            if
                NumCandidates =:= 0 -> error;  %% Contradiction
                NumCandidates =:= 1 ->
                    Digit = get_first_candidate(CandSet),
                    case assign(Grid, Candidates, Row, Col, Digit) of
                        {ok, NewGrid, NewCandidates} ->
                            {ok, NewGrid, NewCandidates, true};  %% Changed
                        error -> error
                    end;
                true ->
                    singleton_elimination(Grid, Candidates, Row, Col + 1)
            end;
        _ ->
            singleton_elimination(Grid, Candidates, Row, Col + 1)
    end.

%% Hidden singles - rows
hidden_singles_rows(Grid, Candidates, 9, _) -> {ok, Grid, Candidates, false};
hidden_singles_rows(Grid, Candidates, Row, 10) ->
    hidden_singles_rows(Grid, Candidates, Row + 1, 1);
hidden_singles_rows(Grid, Candidates, Row, Digit) ->
    {Count, LastCol, AlreadyAssigned} = count_digit_in_row(Grid, Candidates, Row, Digit, 0),
    if
        AlreadyAssigned -> hidden_singles_rows(Grid, Candidates, Row, Digit + 1);
        Count =:= 1 ->
            case assign(Grid, Candidates, Row, LastCol, Digit) of
                {ok, NewGrid, NewCandidates} ->
                    {ok, NewGrid, NewCandidates, true};  %% Changed
                error -> error
            end;
        Count =:= 0 -> error;  %% Contradiction
        true -> hidden_singles_rows(Grid, Candidates, Row, Digit + 1)
    end.

count_digit_in_row(_Grid, _Candidates, _Row, _Digit, Col) when Col =:= 9 ->
    {0, -1, false};
count_digit_in_row(Grid, Candidates, Row, Digit, Col) ->
    case get_grid(Grid, Row, Col) of
        Digit -> {0, -1, true};  %% Already assigned
        0 ->
            CandSet = get_candidates(Candidates, Row, Col),
            HasDig = has_candidate(CandSet, Digit),
            {RestCount, RestLastCol, RestAssigned} = count_digit_in_row(Grid, Candidates, Row, Digit, Col + 1),
            if
                RestAssigned -> {0, -1, true};
                HasDig -> {RestCount + 1, Col, false};
                true -> {RestCount, RestLastCol, false}
            end;
        _ ->
            count_digit_in_row(Grid, Candidates, Row, Digit, Col + 1)
    end.

%% Hidden singles - columns
hidden_singles_cols(Grid, Candidates, 9, _) -> {ok, Grid, Candidates, false};
hidden_singles_cols(Grid, Candidates, Col, 10) ->
    hidden_singles_cols(Grid, Candidates, Col + 1, 1);
hidden_singles_cols(Grid, Candidates, Col, Digit) ->
    {Count, LastRow, AlreadyAssigned} = count_digit_in_col(Grid, Candidates, Col, Digit, 0),
    if
        AlreadyAssigned -> hidden_singles_cols(Grid, Candidates, Col, Digit + 1);
        Count =:= 1 ->
            case assign(Grid, Candidates, LastRow, Col, Digit) of
                {ok, NewGrid, NewCandidates} ->
                    {ok, NewGrid, NewCandidates, true};  %% Changed
                error -> error
            end;
        Count =:= 0 -> error;  %% Contradiction
        true -> hidden_singles_cols(Grid, Candidates, Col, Digit + 1)
    end.

count_digit_in_col(_Grid, _Candidates, _Col, _Digit, Row) when Row =:= 9 ->
    {0, -1, false};
count_digit_in_col(Grid, Candidates, Col, Digit, Row) ->
    case get_grid(Grid, Row, Col) of
        Digit -> {0, -1, true};  %% Already assigned
        0 ->
            CandSet = get_candidates(Candidates, Row, Col),
            HasDig = has_candidate(CandSet, Digit),
            {RestCount, RestLastRow, RestAssigned} = count_digit_in_col(Grid, Candidates, Col, Digit, Row + 1),
            if
                RestAssigned -> {0, -1, true};
                HasDig -> {RestCount + 1, Row, false};
                true -> {RestCount, RestLastRow, false}
            end;
        _ ->
            count_digit_in_col(Grid, Candidates, Col, Digit, Row + 1)
    end.

%% Hidden singles - boxes
hidden_singles_boxes(Grid, Candidates, 9, _) -> {ok, Grid, Candidates, false};
hidden_singles_boxes(Grid, Candidates, BoxIdx, 10) ->
    hidden_singles_boxes(Grid, Candidates, BoxIdx + 1, 1);
hidden_singles_boxes(Grid, Candidates, BoxIdx, Digit) ->
    BoxRow = (BoxIdx div 3) * 3,
    BoxCol = (BoxIdx rem 3) * 3,
    {Count, LastRow, LastCol, AlreadyAssigned} = count_digit_in_box(Grid, Candidates, BoxRow, BoxCol, Digit),
    if
        AlreadyAssigned -> hidden_singles_boxes(Grid, Candidates, BoxIdx, Digit + 1);
        Count =:= 1 ->
            case assign(Grid, Candidates, LastRow, LastCol, Digit) of
                {ok, NewGrid, NewCandidates} ->
                    {ok, NewGrid, NewCandidates, true};  %% Changed
                error -> error
            end;
        Count =:= 0 -> error;  %% Contradiction
        true -> hidden_singles_boxes(Grid, Candidates, BoxIdx, Digit + 1)
    end.

count_digit_in_box(Grid, Candidates, BoxRow, BoxCol, Digit) ->
    Cells = [{R, C} || R <- lists:seq(BoxRow, BoxRow + 2),
                       C <- lists:seq(BoxCol, BoxCol + 2)],
    count_digit_in_cells(Grid, Candidates, Cells, Digit, 0, -1, -1).

count_digit_in_cells(_Grid, _Candidates, [], _Digit, Count, LastR, LastC) ->
    {Count, LastR, LastC, false};
count_digit_in_cells(Grid, Candidates, [{R, C}|Rest], Digit, Count, LastR, LastC) ->
    case get_grid(Grid, R, C) of
        Digit -> {0, -1, -1, true};  %% Already assigned
        0 ->
            CandSet = get_candidates(Candidates, R, C),
            HasDig = has_candidate(CandSet, Digit),
            if
                HasDig -> count_digit_in_cells(Grid, Candidates, Rest, Digit, Count + 1, R, C);
                true -> count_digit_in_cells(Grid, Candidates, Rest, Digit, Count, LastR, LastC)
            end;
        _ ->
            count_digit_in_cells(Grid, Candidates, Rest, Digit, Count, LastR, LastC)
    end.

%% Check if puzzle is solved
is_solved(Grid) ->
    is_solved_check(Grid, 0, 0).

is_solved_check(_, 9, _) -> true;
is_solved_check(Grid, Row, 9) ->
    is_solved_check(Grid, Row + 1, 0);
is_solved_check(Grid, Row, Col) ->
    case get_grid(Grid, Row, Col) of
        0 -> false;
        _ -> is_solved_check(Grid, Row, Col + 1)
    end.

%% Find cell with minimum remaining values (MRV heuristic)
find_mrv_cell(Grid, Candidates) ->
    find_mrv_cell(Grid, Candidates, 0, 0, none, 10).

find_mrv_cell(_, _, 9, _, Best, _) -> Best;
find_mrv_cell(Grid, Candidates, Row, 9, Best, MinCount) ->
    find_mrv_cell(Grid, Candidates, Row + 1, 0, Best, MinCount);
find_mrv_cell(Grid, Candidates, Row, Col, Best, MinCount) ->
    case get_grid(Grid, Row, Col) of
        0 ->
            CandSet = get_candidates(Candidates, Row, Col),
            Count = count_candidates(CandSet),
            if
                Count < MinCount ->
                    find_mrv_cell(Grid, Candidates, Row, Col + 1, {Row, Col}, Count);
                true ->
                    find_mrv_cell(Grid, Candidates, Row, Col + 1, Best, MinCount)
            end;
        _ ->
            find_mrv_cell(Grid, Candidates, Row, Col + 1, Best, MinCount)
    end.

%% Search with backtracking
search(Grid, Candidates) ->
    case is_solved(Grid) of
        true -> {ok, Grid};
        false ->
            %% Propagate constraints
            case propagate(Grid, Candidates) of
                {ok, NewGrid, NewCandidates} ->
                    case is_solved(NewGrid) of
                        true -> {ok, NewGrid};
                        false ->
                            %% Find cell with MRV
                            case find_mrv_cell(NewGrid, NewCandidates) of
                                none -> error;  %% No empty cells but not solved (contradiction)
                                {Row, Col} ->
                                    CandSet = get_candidates(NewCandidates, Row, Col),
                                    try_candidates(NewGrid, NewCandidates, Row, Col, CandSet, 1)
                            end
                    end;
                error -> error
            end
    end.

try_candidates(_Grid, _Candidates, _Row, _Col, _CandSet, 10) -> error;
try_candidates(Grid, Candidates, Row, Col, CandSet, Digit) ->
    case has_candidate(CandSet, Digit) of
        true ->
            %% Try this candidate
            case assign(Grid, Candidates, Row, Col, Digit) of
                {ok, NewGrid, NewCandidates} ->
                    case search(NewGrid, NewCandidates) of
                        {ok, Solution} -> {ok, Solution};
                        error ->
                            %% Backtrack
                            try_candidates(Grid, Candidates, Row, Col, CandSet, Digit + 1)
                    end;
                error ->
                    %% Assignment caused contradiction
                    try_candidates(Grid, Candidates, Row, Col, CandSet, Digit + 1)
            end;
        false ->
            try_candidates(Grid, Candidates, Row, Col, CandSet, Digit + 1)
    end.

%% Solve CP
solve_cp(Grid, Candidates) ->
    search(Grid, Candidates).
